src/esp32/board.rs

```diff
@@ -1,53 +1,228 @@
+    /// Down the road we will need to wrap the Esp32Board in a singleton instance owning the peripherals and giving them as requested.
+    /// The potential approach is described in esp32/motor.rs:383
+    pub(crate) fn from_config(cfg: ConfigType) -> anyhow::Result<BoardType> {
+        let (analogs, pins) = match cfg {
+            ConfigType::Static(cfg) => {
+                let analogs = if let Ok(analogs) =
+                    cfg.get_attribute::<Vec<AnalogReaderConfig>>("analogs")
+                {
+                    let analogs: Vec<Rc<RefCell<dyn AnalogReader<u16, Error = anyhow::Error>>>> =
+                        analogs
+                            .iter()
+                            .filter_map(|v| {
+                                let adc1 = Rc::new(RefCell::new(
+                                    AdcDriver::new(
+                                        unsafe { ADC1::new() },
+                                        &Config::new().calibration(true),
+                                    )
+                                    .ok()?,
+                                ));
+                                let chan: Rc<
+                                    RefCell<dyn AnalogReader<u16, Error = anyhow::Error>>,
+                                > = match v.pin {
+                                    32 => {
+                                        let p: Rc<
+                                            RefCell<dyn AnalogReader<u16, Error = anyhow::Error>>,
+                                        > = Rc::new(RefCell::new(Esp32AnalogReader::new(
+                                            v.name.to_string(),
+                                            AdcChannelDriver::<_, Atten11dB<ADC1>>::new(unsafe {
+                                                esp_idf_hal::gpio::Gpio32::new()
+                                            })
+                                            .ok()?,
+                                            adc1,
+                                        )));
+                                        Some(p)
+                                    }
+                                    33 => {
+                                        let p: Rc<
+                                            RefCell<dyn AnalogReader<u16, Error = anyhow::Error>>,
+                                        > = Rc::new(RefCell::new(Esp32AnalogReader::new(
+                                            v.name.to_string(),
+                                            AdcChannelDriver::<_, Atten11dB<ADC1>>::new(unsafe {
+                                                esp_idf_hal::gpio::Gpio33::new()
+                                            })
+                                            .ok()?,
+                                            adc1,
+                                        )));
+                                        Some(p)
```

Comment:
It looks like AdcChannelDriver::new only needs an argument of type `impl Peripheral<P = T> + 'd`. If you made a helper function that constructed `esp_idf_hal::gpio::GpioN` by number, could you just match on the range to validate that you had an ADC capable pin and then eliminate the duplication?

Reply:
The ADC trait is not implemented for `AnyOutputPin` onlye for GpioXX by the `esp-idf-hal` crate. We can't construct a GpioN just using the pin number. I am leaning toward moving all this logic on the esp32 board and use a macro to generate the relevant code. But that would take a separate PR